### PR TITLE
Support loading distributed checkpoints for FSDP in Trainer

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part 2 (https://github.com/Lightning-AI/lightning/pull/18364) completes the support for sharded / distributed checkpointing with FSDP in Trainer. The FSDP user guide will be updated in a follow-up PR. 

There are minor breaking changes (when using FSDP only):
- The checkpoint gets loaded after `configure_sharded_model()`, that is, after the model is sharded and put on the GPU. This is to avoid expensively loading the checkpoint in CPU memory before sharding (the model may not fit!). The other strategy that also does this is DeepSpeed. This change is necessary now to support sharded checkpoints, but will also be needed for supporting meta device later (also for full-state checkpoints).
- The `FSDPStrategy.load_optimizer_state_dict` method becomes a no-op. 
- The `FSDPStrategy.load_model_state_dict` method becomes a no-op. 

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/pytorch/strategies/fsdp.py [hunk 1]
```diff
@@ -148,7 +151,7 @@ def __init__(
         activation_checkpointing: Optional[Union[Type[Module], List[Type[Module]]]] = None,
         activation_checkpointing_policy: Optional["_POLICY"] = None,
         sharding_strategy: "_SHARDING_STRATEGY" = "FULL_SHARD",
-        state_dict_type: Literal["full", "sharded"] = "full",
+        state_dict_type: Literal["full", "sharded"] = "full",  # FIXME: make sharded the default
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/fsdp.py:154]: Do we want to make this change yes or no?

Pros:
- consistent with fabric
Cons:
- breaking change

Comment (carmocca) [reply src/lightning/pytorch/strategies/fsdp.py:154]: I prefer "full" as the default in general since it's easier to deal with for most users.

File: src/lightning/pytorch/strategies/fsdp.py [hunk 2]
```diff
@@ -451,6 +464,10 @@ def lightning_module_state_dict(self) -> Dict[str, Any]:
         with state_dict_ctx:
             return self.model.state_dict()
 
+    def load_model_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
+        # Override to do nothing, the FSDP already loaded the states in `load_checkpoint()`
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/fsdp.py:468]:
```suggestion
        # Override to do nothing, FSDP already loaded the states in `load_checkpoint()`
```

**Review by carmocca — APPROVED** [review]
LGTM! It looks just as in Fabric now
